test(tiering): skip tests if O_DIRECT is unsupported

glevkovich opened the pull request:
fix(tiering): support hosts without O_DIRECT

Enable running the full test suite on hosts lacking O_DIRECT support
(e.g., laptops, encrypted /home) by automatically falling back to
standard buffered I/O.

Previously, these tests would fail or required manual exclusion on
filesystems that do not support Direct I/O.

Changes:
* Add CMake probe to detect O_DIRECT support on the build host.
* Conditionally define O_DIRECT flags in disk_storage and
  snapshot_storage.
* If unsupported, tests execute using standard page cache IO instead
  of being skipped.

dranikpg commented on CMakeLists.txt:
```diff
@@ -52,6 +52,33 @@ set(CMAKE_REQUIRED_FLAGS "-fsanitize=undefined")
+  add_compile_definitions(DIRECT_IO_SUPPORTED) # To be able to filter out test cases and not the whole test suite
+else()
+  message(STATUS "Build host does not support O_DIRECT (note: this check only validates the build host, target capabilities may differ when cross-compiling) - relevant tiering tests will be disabled")
+endif()
```
Not sure if we need it automatic, if this was a manual flag you could ifdef out O_DIRECT from at least snapshot storage and it would've (should've at least) worked

glevkovich replied:
Thanks for the suggestion. I’ve updated the PR to handle this automatically, but with a better approach: CMake now probes the build host for O_DIRECT support. If it's not available, the code gracefully falls back to standard buffered I/O.

This ensures that all tests run (instead of being skipped), maintaining functional coverage even on systems without Direct I/O.

Note: There is similar technical debt in helio/io/file.cc, but since all tests are currently passing for me without touching it, I plan to address that in a separate PR later.

glevkovich committed:
fix(tiering): support hosts without O_DIRECT

Enable running the full test suite on hosts lacking O_DIRECT support
(e.g., laptops, encrypted /home) by automatically falling back to
standard buffered I/O.

Previously, these tests would fail or required manual exclusion on
filesystems that do not support Direct I/O.

Changes:
* Add CMake probe to detect O_DIRECT support on the build host.
* Conditionally define O_DIRECT flags in disk_storage and
  snapshot_storage.
* If unsupported, tests execute using standard page cache IO instead
  of being skipped.

Signed-off-by: Gil Levkovich <69595609+glevkovich@users.noreply.github.com>

glevkovich commented on CMakeLists.txt:
```diff
@@ -52,6 +52,33 @@ set(CMAKE_REQUIRED_FLAGS "-fsanitize=undefined")
+    int fd = open(\"direct_io_probe.tmp\", O_CREAT | O_WRONLY | O_DIRECT, 0666);
+    if (fd < 0) return 1;
+    close(fd);
+    unlink(\"direct_io_probe.tmp\");
```
We dont need all this, what I did is the absolute minimal and it works.

glevkovich commented:
Closing this PR for now, as this requires more work than I thought and it's too riskt at this stage. I'll be back to it later on with a better solution.
